src/dotnet/commands/dotnet-build/CompileContext.cs

```diff
@@ -54,32 +54,17 @@ public bool Compile(bool incremental)
         {
             CreateOutputDirectories();
 
-            // compile dependencies
-            foreach (var dependency in Sort(_dependencies.ProjectDependenciesWithSources))
-            {
-                if (incremental)
-                {
-                    var dependencyProjectContext = ProjectContext.Create(dependency.Path, dependency.Framework);
-
-                    if (!DependencyNeedsRebuilding(dependencyProjectContext, new ProjectDependenciesFacade(dependencyProjectContext, _args.ConfigValue)))
-                    {
-                        continue;
-                    }
-                }
-
-                if (!InvokeCompileOnDependency(dependency))
-                {
-                    return false;
-                }
-            }
+            return CompileDendencies(incremental) && CompileRootProject(incremental);
```

Comment:
It's been nagging me ever since first wrote that code :)
I put a TODO somewhere in there about potentially introducing an explicit DAG.

The main complication is that behaviour differs ever so slightly for the root project only (different args sent to compile, special treatment due to --no-dependency flag, etc). I am waiting for a couple of more features / changes to see which solution should be best. Right now I am going towards DAG with RootProjectNode and DependencyNode. The abstract node can have methods like NeedsRebuilding.

But yeah, near future :)